File: answers/other/answers/lintcode/Longest Palindromic Substring.cpp
class Solution {
public:
    /**
     * @param s input string
     * @return the longest palindromic substring
     */
    string longestPalindrome(string& s) {
        // Write your code here
        std::vector<std::vector<bool> > dp(s.length(), std::vector<bool> (s.length(), false));
        size_t left = 0;
        size_t length = 0;
        for (size_t i = 0; i < s.length(); ++ i) {
            for (size_t j = 0; i + j < s.length(); ++ j) {
                dp[i][j] = 0 == i || (s[j] == s[j + i] && (i <= 1 || dp[i - 2][j + 1]));
                if (dp[i][j] && i > length) {
                    left = j;
                    length = i;
                }
            }
        }
        return s.substr(left, length + 1);
    }
};

File: answers/leetcode/Longest Palindromic Substring/Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        int left = 0;
        int right = 0;

        for (int i = 0; i < s.length(); i++) {
            int j = 0;
            while (i - j >= 0 && i + j < s.length() && s.charAt(i - j) == s.charAt(i + j)) {
                j++;
            }
            if (j * 2 - 1 > right - left + 1) {
                left = i - j + 1;
                right = i + j - 1;
            }

            j = 0;
            while (i - j >= 0 && i + j + 1 < s.length() && s.charAt(i - j) == s.charAt(i + j + 1)) {
                j++;
            }
            if (j * 2 > right - left + 1) {
                left = i - j + 1;
                right = i + j;
            }
        }

        return s.substring(left, right + 1);
    }
}
//Runtime: 49 ms, faster than 48.66% of Java online submissions for Longest Palindromic Substring.
//Memory Usage: 39.9 MB, less than 45.36% of Java online submissions for Longest Palindromic Substring.
